api/diff.go

```diff
@@ -45,18 +45,46 @@ func handleAPIDiffGet(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	var beforeAndAfter shared.ProductSpecs
-	beforeAndAfter, err = shared.ParseBeforeAndAfterParams(r)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+	var runs shared.TestRuns
+	if len(runIDs) > 0 {
+		runs, err = runIDs.LoadTestRuns(ctx)
+		if err != nil {
+			if multiError, ok := err.(appengine.MultiError); ok {
+				all404s := true
+				for _, err := range multiError {
+					if err != datastore.ErrNoSuchEntity {
+						all404s = false
+					}
+				}
+				if all404s {
+					http.NotFound(w, r)
+					return
+				}
+			}
+		}
+	} else {
+		// NOTE: We use the same params as /results, but also support
+		// 'before' and 'after' and 'filter'.
+		runFilter, err := shared.ParseTestRunFilterParams(r)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		var beforeAndAfter shared.ProductSpecs
+		beforeAndAfter, err = shared.ParseBeforeAndAfterParams(r)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if len(beforeAndAfter) > 0 {
+			runFilter.Products = beforeAndAfter
+		}
+		runs, err = LoadTestRunsForFilters(ctx, runFilter)
 	}
 
-	if len(beforeAndAfter) > 0 {
-		runFilter.Products = beforeAndAfter
-	}
-	var runs shared.TestRuns
-	if runs, err = LoadTestRunsForFilters(ctx, runFilter); err != nil {
+	if err != nil {
```

Comment:
nit: Let's move this up to the end of the `if len(runIDs) > 0 { ... }` block, since all `err` statements in its `else` use `..., err := ` (living inside the `else`'s block scope). Just this block can move; IIUC, the `else if len(runs) != 2` can become ` len(runs) != 2`.

Reply:
Renamed the local vars to `parseErr` to be more readable.